R/predict.R

```diff
@@ -0,0 +1,512 @@
+    desired_dates <- as.Date(
+      c(na_pad, desired_dates, na_pad)
+    )
+  }
+  format_newdata_dataframe(
+    fit = object,
+    parameter = parameter,
+    desired_dates = desired_dates,
+    timesteps = timesteps
+  )
+}
+
+#' Format predicted data frame with uniform minimal specification
+#' @noRd
+format_predicted_dataframe <- function(
+    parameter,
+    fitted,
+    newdata,
+    delta,
+    gi_pmf) {
+  if (parameter == "r") {
+    # Difference calculation for `r` parameter
+    timestep_first_row <- which((fitted[[".row"]] - 1) %% 2 == 0)
+    fitted <- data.frame(
+      .row = (fitted[timestep_first_row, ".row"] + 1) / 2,
+      .response = discrete_diff_derivative(fitted[[".fitted"]]),
+      .draw = fitted[timestep_first_row, ".draw"]
+    )
+  } else if (parameter == "Rt") {
+    # Rt calculation
+    fitted <- rt_by_group(fitted,
+      group_cols = ".draw",
+      value_col = ".fitted",
+      vec = gi_pmf
+    )
+  }
+
+  # Merge with newdata and select required columns
+  merged <- merge(fitted,
+    newdata,
+    by = ".row"
+  )[, c(
+    "reference_date",
+    ".response",
+    ".draw"
+  )]
+  merged$reference_date <- as.Date(merged$reference_date)
+  merged[which(!is.na(merged[["reference_date"]])), ]
+}
+
+#' Turn input parameters into a dataframe to pass to model
+#' @noRd
+format_newdata_dataframe <- function(
+    fit,
+    parameter,
+    desired_dates,
+    timesteps,
+    gi_pmf,
+    delta) {
+  newdata <- gratia::data_slice(fit[["model"]], timestep = timesteps)
+
+  if (parameter == "r") {
+    # Join key when adding metadata to posterior sim. Setting
+    # every other row to NA ensures we get one row per date-draw.
+    newdata[".row"] <- interleave(seq(1, length(timesteps) / 2, 1), NA)
+  } else {
+    newdata[".row"] <- seq_along(timesteps)
+  }
+  newdata["reference_date"] <- desired_dates
+  return(newdata)
+}
+
+#' Centered difference derivative.
+#' @noRd
+discrete_diff_derivative <- function(vals) {
+  len <- length(vals)
+  t0 <- seq(1, len - 1, 2)
+  t1 <- seq(2, len, 2)
+
+  (vals[t1] - vals[t0]) / 2
+}
+
+#' Generate size in timesteps of centered two day window
+#' @noRd
+compute_delta <- function(fit) {
+  min_date <- fit[["min_date"]]
+  timesteps <- dates_to_timesteps(
+    c(
+      min_date,
+      min_date + 1
+    ),
+    min_supplied_date = min_date,
+    max_supplied_date = fit[["max_date"]]
+  )
+  return(timesteps[2] - timesteps[1])
+}
+
+#' Convolve by draw. In the future, additional groups could be detected and
+#' specified here.
+#' @noRd
+rt_by_group <- function(df, group_cols, value_col, vec) {
+  df[["group_id"]] <- interaction(df[group_cols], drop = TRUE)
+  unique_groups <- as.character(unique(df[["group_id"]]))
+
+  grouped_rt <- lapply(unique_groups, function(group_id) {
+    group_df <- df[df[["group_id"]] == group_id, ]
+    # Produces vec w/ length nrow(group_df) + length(vec) - 1
+    convolved <- stats::convolve(group_df[[value_col]], rev(vec), type = "open")
```

Comment:
This requires you to use a zero on the first pmf to get the right renewal? Are you doing so in the example you should in the PR?

Reply:
🤦 duh

![test](https://github.com/user-attachments/assets/060661b1-965c-4532-a82a-d015ad349106)

Will fix the PMF generation in `data_raw/` in #69